### PR TITLE
Bring back portal from master to patch 4

### PR DESCRIPTION
**Pull request by sjpadgett**
#### Short description of what this resolves:
Porting new portal features to patch 4
This PR replaces PR #4901

### CODE REVIEW

**Comment by sjpadgett** [comment]
Still testing from new install through patch 4.

**Comment by sjpadgett** [comment]
Setup new install, patched up to patch3 then applied current patch 4.
Testing well. @bradymiller I'm comfortable with bringing in if you need or can wait another day.

File: portal/home.php [hunk 1]
```diff
@@ -50,7 +52,8 @@
     }
 }
 if ($newcnt > 0 && $_SESSION['portal_init']) {
-    $whereto = $_SESSION['whereto'] = '#secure-msgs-card';
+    $whereto = '#secure-msgs-card';
+    OpenEMR\Common\Session\SessionUtil::setSession('whereto', '#secure-msgs-card');
```
Comment (bradymiller) [inline portal/home.php:56]: this doesn't look right unless using this outside the portal for core openemr session. in portal can set sessions (there is no write block) and when use this setSession it will actually create a core (OpenEMR) session.

File: portal/add_edit_event_user.php [hunk 2]
```diff
@@ -487,8 +487,9 @@
     $user = sqlQueryNoLog("SELECT users.username FROM users WHERE authorized = 1 And id = ?", array($_POST['form_provider_ae']));
     $rtn = addPnote($_POST['form_pid'], $note, 1, 1, $title, $user['username'], '', 'New');
 
-    $_SESSION['whereto'] = 'appointmentcard';
-    header('Location:./home.php#appointmentpanel');
+
+    OpenEMR\Common\Session\SessionUtil::setSession('whereto', '#appointmentcard');
```
Comment (bradymiller) [inline portal/add_edit_event_user.php:491]: same issue as noted below. if these setSession calls are meant for use in portal, then would change these all to direct session changes or else will be creating a new core OpenEMR session:
```
$_SESSION['whereto'] = '#appointmentcard';
```

File: portal/lib/paylib.php [hunk 3]
```diff
@@ -37,7 +37,7 @@
 use OpenEMR\Common\Crypto\CryptoGen;
 
 if ($_SESSION['portal_init'] !== true) {
-    $_SESSION['whereto'] = '#paymentcard';
+    OpenEMR\Common\Session\SessionUtil::setSession('whereto', '#paymentcard');
```
Comment (bradymiller) [inline portal/lib/paylib.php:40]: This is another place where will be creating a Openemr core session inside the portal with this function. This script works in both openemr and portal and performance not an issue. so would directly set the session variables (ie. go back to line that was removed) and then use flag prior to the globals.php include that turns off the session lock write so direct setting of session will also work when this script is used in openemr core.

File: portal/lib/persist.php [hunk 4]
```diff
@@ -38,8 +38,8 @@
 }
 
 if (!empty($data['where'] ?? null)) {
-    $_SESSION['whereto'] = $data['where'];
+    OpenEMR\Common\Session\SessionUtil::setSession('whereto', $data['where']);
```
Comment (bradymiller) [inline portal/lib/persist.php:41]: same issue as i above paylib.php comment

File: sql/patch.sql [hunk 5]
```diff
@@ -147,3 +147,106 @@ KEY (`success`)
...
+#EndIf
+
+SET @ai_exist =(SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = 'document_templates' AND COLUMN_NAME = 'id' AND 'EXTRA' LIKE '%auto_increment%' LIMIT 1);
+SET @query = IF (@ai_exist = 0, 'ALTER TABLE `document_templates` CHANGE `id` `id` BIGINT(21) UNSIGNED NOT NULL AUTO_INCREMENT', '');
+PREPARE statement FROM @query;
+EXECUTE statement;
```
Comment (bradymiller) [inline sql/patch.sql:185]: can't this change just be made directly in the above document_templates table definition?

File: sql/patch.sql [hunk 6]
```diff
@@ -147,3 +147,106 @@ KEY (`success`)
...
+INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`) VALUES ('Patient_Groupings','group_3','Group III',30,0,0);
+INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`) VALUES ('Patient_Groupings','group_4','Group IV',40,0,0);
+INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`) VALUES ('Patient_Groupings','group_5','Group V',50,0,0);
+#EndIf
+
+#IfMissingColumn document_templates send_date
+ALTER TABLE `document_templates` CHANGE `exclude_portal` `send_date` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP, CHANGE `exclude_dashboard` `end_date` DATETIME DEFAULT NULL;
+ALTER TABLE `document_templates` CHANGE `profile` `profile` VARCHAR(63) NOT NULL, CHANGE `category` `category` VARCHAR(63) NOT NULL;
+ALTER TABLE `document_templates` DROP INDEX `location`, ADD UNIQUE `location` (`pid`, `profile`, `category`, `template_name`);
+#EndIf
```
Comment (bradymiller) [inline sql/patch.sql:234]: can't this change just be made directly in the above document_templates table schema?

File: sql/patch.sql [hunk 7]
```diff
@@ -147,3 +147,106 @@ KEY (`success`)
...
+
+#IfMissingColumn document_template_profiles member_of
+ALTER TABLE `document_template_profiles` ADD `member_of` VARCHAR(64) NOT NULL;
+ALTER TABLE `document_template_profiles` ADD `active` TINYINT(1) NOT NULL DEFAULT '0';
+ALTER TABLE `document_template_profiles` ADD `recurring` TINYINT(1) NOT NULL DEFAULT '0',ADD `event_trigger` VARCHAR(31) NOT NULL, ADD `period` INT(4) NOT NULL;
+ALTER TABLE `document_template_profiles` CHANGE `profile` `profile` VARCHAR(64) NOT NULL, CHANGE `category` `category` VARCHAR(64) NOT NULL, CHANGE `template_name` `template_name` VARCHAR(255) NOT NULL;
+ALTER TABLE `document_template_profiles` DROP INDEX `location`, ADD UNIQUE `location` (`profile`, `template_id`, `member_of`);
+#EndIf
```
Comment (bradymiller) [inline sql/patch.sql:242]: can't this change just be made directly in the above document_template_profiles table schema?

File: sql/patch.sql [hunk 8]
```diff
@@ -147,3 +147,106 @@ KEY (`success`)
 ) ENGINE=InnoDB;
 #EndIf
```
Comment (bradymiller) [inline sql/patch.sql:149]: for the document_templates and document_template_profiles could simplify by just using same tables from database.sql (assuming those are the most updated tables) and not need to then also then modify them later on in this script

**Comment by bradymiller** [comment]
code review done (didn't test). just couple issues noted.
1. couple database issues
2. setSession issue

File: portal/home.php (hunk 1 above)
Comment (sjpadgett) [reply portal/home.php:56]: @bradymiller unsure why I changed this. maybe a senior moment or I was tracking a bug. still, can't be right either way so thanks.

File: sql/patch.sql (hunk 5 above)
Comment (sjpadgett) [reply sql/patch.sql:185]: yes that'd make sense however, I have to track all the pre testing patches I have in the wild so treat this as upgrade...

File: portal/home.php (hunk 1 above)
Comment (sjpadgett) [reply portal/home.php:56]: yeah, this all started going back to when I converted the portal to a single page app plus, portal scripts are shared between just a portal session and an openemr session for dashboard.
Brady separated these out when he did the new session manager classes. You're right we had an issue with persisting where patient last viewed across refresh however, I fixed by adding a persist function on menu clicks. But session wasn't being saved correctly thus I added the sessionWrite() which made it work not realizing I was writing to an openemr session. Weird thing with php sessions(not cookies) the whereto session writes still appear in the $_SESSION super as if the same cookie!

Too much for my fragile brain!:)

In the end I had a bug in my persist function where the whereto was immediately being overwritten which caused me to embark on the sessionWrite changes in the first place.

So long story short, this is Brady's fault is what I'm going with.:)

**Comment by stephenwaite** [comment]
Wow!
southpark_cheering
